cdap-ui/app/cdap/components/AbstractWidget/SqlConditionsWidget/index.tsx

```diff
@@ -0,0 +1,270 @@
+      } catch (e) {
+        mapInputSchema[input.name] = [];
+        error = 'Error parsing input schemas.';
+      }
+    });
+    if (stageList.length < 2) {
+      error = 'Please connect 2 or more stages.';
+    }
+    return { error, stageList, mapInputSchema };
+  };
+
+  private init = () => {
+    const { error, stageList, mapInputSchema } = this.initializeOptions();
+    let warning = null;
+    let rules = [];
+    if (!this.props.value) {
+      this.setState({ error, stageList, mapInputSchema }, this.addRule);
+      return;
+    }
+    const modelSplit = this.props.value.split('&').map((rule) => rule.trim());
+    modelSplit.forEach((rule) => {
+      const rulesArr = [];
+      rule.split('=').forEach((field) => {
+        const splitField = field.trim().split('.');
+        // Not including rule if stage has been disconnected
+        if (stageList.indexOf(splitField[0]) === -1) {
+          return;
+        }
+        rulesArr.push({
+          stageName: splitField[0],
+          fieldName: splitField[1],
+        });
+      });
+      // Missed fields scenario will happen if the user connects more stages into the join node
+      // after they have configured join conditions previously
+      const missedFields = stageList.filter((stage) => {
+        const filteredRule = rulesArr.filter((field) => {
+          return field.stageName === stage;
+        });
+        return filteredRule.length === 0 ? true : false;
+      });
+      if (missedFields.length > 0) {
+        missedFields.forEach((field) => {
+          rulesArr.push({
+            stageName: field,
+            fieldName: mapInputSchema[field][0],
+          });
+        });
+        warning =
+          "Input stages have changed since the last time you edit this node's configuration. Please verify the condition is still valid.";
+      }
+      rules = [...rules, rulesArr];
+    });
+    this.setState({ error, warning, rules, stageList, mapInputSchema }, this.formatOutput);
+  };
+
+  public render() {
+    const { classes, disabled } = this.props;
+    return (
+      <div className={classes.root}>
+        <If condition={this.state.error && this.state.stageList.length > 0}>
+          <div className={classnames(classes.textDanger, 'text-danger')}>{this.state.error}</div>
+        </If>
+        <If condition={this.state.warning && !this.state.error}>
+          <div className="text-warning">{this.state.warning}</div>
+        </If>
+        <div className={classes.rulesContainer}>
+          {this.state.rules.map((rule, i) => {
+            return (
+              <Rule
+                rule={rule}
+                ruleIdx={i}
+                rulesCount={this.state.rules.length}
+                inputSchema={this.state.mapInputSchema}
+                addRule={this.addRule}
+                disabled={disabled}
+                updateRule={(changedRule) => this.updateRule(changedRule, i)}
+                deleteRule={this.deleteRule}
```

Comment:
`key` prop

Reply:
defaults to using index as key. We're mapping over array of array of objects. We don't have a hash function to generate unique id.